docs/design/rosbag2_playback_time.md

```diff
@@ -0,0 +1,200 @@
+ * Internally, it may own an rclcpp::Clock, but does not override the class in order to implement
+ * a slightly different API.
+ */
+class PlayerClock
+{
+public:
+  /**
+   * Constructor.
+   *
+   * \param use_sim_time
+   *   If false, starts paused.
+   *     Use member functions to configure before calling set_paused(false)
+   *   If true, subscribe to /clock to provide time
+   *     All time control (rate, pause, jump) is disabled, and will not publish to /clock
+   */
+  explicit PlayerClock(bool use_sim_time);
+
+  virtual ~PlayerClock();
+
+  // rclcpp::Clock interface
+  /**
+   * Provides the current time according to the clock's internal model.
+   *  if use_sim_time: provides current ROS Time (with optional extrapolation - see "Clock Rate and Time Extrapolation" section)
+   *  if !use_sim_time: calculates current "Player Time" based on starting time, playback rate, pause state.
+   *    this means that /clock time will match with the recorded messages time, as if we are fully reliving the recorded session
+   */
+  rclcpp::Time now();
+
+  /**
+   * Sleeps (non-busy wait) the current thread until the provided time is reached - according to this Clock
+   *
+   * If time is paused, the requested time may never be reached
+   * `real_time_timeout` uses the internal steady clock, if the timeout elapses, return false
+   * If jump() is called, return false, allowing the caller to handle the new time
+   * Return true if the time has been reached
+   */
+  bool sleep_until(rclcpp::Time until, rclcpp::Duration real_time_timeout);
+
+  // Time Control interface
+
+  /**
+   * Pauses/resumes time. Defaults to true (paused)
+   * While paused, `now()` will repeatedly return the same time, until resumed
+   *
+   * Note: this could have been implemented as `set_rate(0)`, but this interface allows this clock to maintain the
+   * clock's rate internally, so that the caller does not have to save it in order to resume.
+   */
+  void set_paused(bool paused);
+  bool get_paused() const;
+
+  /**
+    * Changes the rate of playback. Defaults to 1.0
+    * \param rate: must be nonzero positive. To pause playback, use \sa set_paused instead
+    * \throws std::invalid_argument if rate is <= 0.0
+    */
+  void set_rate(float rate);
+  float get_rate() const;
+
+  /**
+   * Set the rate that the clock will be published. Defaults to 0.
+   * \param frequency If this is set to <= 0, then the clock will not be published
+   */
+  void set_clock_publish_frequency(float frequency);
+  float get_clock_publish_frequency() const;
+
+  /**
+    * Change the current internally maintained offset so that next published time is different.
+    *
+    * This will trigger any registered JumpHandler callbacks.
+    * Call this with the first message timestamp for a bag before starting playback (otherwise this will return current wall time)
+    */
+  void jump(rclcpp::Time time);
+
+  rclcpp::JumpHandler::SharedPtr
+  create_jump_callback(
```

Comment:
It is unclear from the documentation if these callbacks will be executed if use_sim_time == true and the incoming sim time jumps back